### ISSUE TITLE
Fails to build on the pkg-build server

### ISSUE THREAD

**Issue by samth** [emina/rosette#17]
Unfortunately, right now the Racket automatic pkg-build server for Racket fails to build Rosette, because it doesn't allow it to download Z3. You can see the current error here: https://plt.eecs.northwestern.edu/pkg-build/server/built/fail/rosette.txt

This means that Rosette's documentation doesn't appear on https://docs.racket-lang.org among other issues. 

Probably the right fix is to enable the docs to run without actually using z3. I'm happy to help with that if needed.

**Comment by emina** [comment]
Thanks for pointing this out Sam! I didn't notice the failure, since we were able to install Rosette through the package manager on several machines / OSs.

Is there an idiomatic way to fake `examples` and `interaction` results in Scribble?  If not, I'll hack up something (probably by precomputing the results from Rosette and dumping them into a file that will be read by Scribble).  Better ideas are most welcome :)

**Comment by samth** [comment]
There is such a library, which I can't remember the name of, by @rmculpepper. I'll see if I can find it a bit later.

**Comment by emina** [comment]
Cool!  A library would definitely make this easier.

**Comment by rmculpepper** [comment]
Yes, it's [`make-log-based-eval`](http://docs.racket-lang.org/scribble/eval.html#%28def._%28%28lib._scribble%2Feval..rkt%29._make-log-based-eval%29%29) in `scribble/example`. It requires the results to be serializable, but it just occurred to me that I could fix that if it would help.

**Comment by emina** [comment]
Is it possible to use `make-log-based-eval` with `rosette`?  If I'm understanding the docs correctly, the resulting evaluator uses `racket/base` as the language.  Thanks!

**Comment by bennn** [comment]
Try this (maybe there's a better way, but seems to work):
```
```

**Comment by emina** [comment]
@bennn Do you mean that the resulting evaluator won't work, or that it will work but not output results correctly to the log file?

**Comment by rmculpepper** [comment]
You can probably just require the rosette bindings over the racket ones. Try this:

    @(define my-eval (make-log-based-eval "file.log" 'record))
    @(my-eval '(require rosette))

I can add an option to `make-log-based-eval` that lets you provide your own evaluator to wrap.

**Comment by emina** [comment]
Thanks @benn and @rmculpepper!  I think I'll try changing `current-eval`.  I'll let you know how it goes.

**Comment by emina** [comment]
A quick update:  I've modified to docs to build without invoking Z3.  Documentation now builds on the server!  The installation continues to be marked as failing, however, because the server can't download Z3.  If you have any suggestions for fixing the latter, I'd be happy to try them out.  Thanks all for the pointers!

**Comment by bennn** [comment]
Does the server need z3? If not you could just catch the `exn:fail:network?` exception & print a message.

**Comment by emina** [comment]
The server shouldn't need Z3 ...  I updated the code to catch any failure to install Z3 and print a message.  Let's see if that helps!

**Comment by emina** [comment]
That didn't seem to help.  The [error log](http://pkg-build.racket-lang.org/server/built/fail/rosette.txt) seems to indicate that the error is not being thrown inside of the installer (so catching the exception there didn't help).  Do you know if it is possible to catch the error?

**Comment by samth** [comment]
I think that error log is from before your most recent commit -- the builds only happen once a day, and the date stamp says `6:27:03am`.

**Comment by emina** [comment]
I see.  So, even though the server pulls more frequently than once a day, it only builds once a day?  I assumed that pull => build.  When I click on the 'rosette' link, the pop-up windows shows the latest checksum (70ba0ef07ae74820f1c16ce757d1ae18b141bca5), and the 'updated' field says 11:20:50 AM PDT.

**Comment by samth** [comment]
Right, the pkg server updates every hour, so if I do `raco pkg update rosette` I get the latest version. The pkg-build server is separate (there's actually a second one, which builds using a snapshot build) and runs every night.

**Comment by emina** [comment]
Got it.  We'll see tomorrow if it worked!

**Comment by samth** [comment]
It now builds successfully! Unfortunately lots of the tests fail, since there's no z3, but that's probably expected: http://pkg-build.racket-lang.org/server/built/test-fail/rosette.txt

There are also some missing dependencies: http://pkg-build.racket-lang.org/server/built/deps/rosette.txt

**Comment by emina** [comment]
Yay!

Is there a way to prevent the package server from running the tests?  Many / most of them need Z3 to run successfully.  

Dependencies should be easier to handle.  I take it the solution is to declare them in the top-level [info.rkt](https://github.com/emina/rosette/blob/master/info.rkt) file as 'deps' (for r6rs-lib, rackunit-lib, slideshow-lib) and 'build-deps' (for pict-doc)?

**Comment by samth** [comment]
For dependencies, you're right on. Note that `raco setup --check-pkg-deps --fix-pkg-deps rosette` will fix them automatically for you.

For tests, there isn't a way to disable them specifically for the pkg server. There are a few options here (in roughly the order that seems best to me).
- check for the presence of z3 and skip the tests
- check for the environment variable `PLT_PKG_BUILD_SERVICE` and skip tests if it is set
- leave them failing

**Comment by emina** [comment]
Cool, that was an easy fix.  Great automation! :)

I don't like the failing tests either, but the requirement for a working copy of Z3 is pretty pervasive.  The code base would have to include a lot of these special-case checks, which seems ugly.  I think I'll leave this be for now, until / unless we can think of a more centralized solution (preferably a single switch / check somewhere).

**Comment by samth** [comment]
Thinking about what the right switch would be ... @mflatt, would an `info.rkt` field like `pkg-build-test-omit-paths` seem sensible? This would require some way to communicate that to `raco test` as well, but that seems possible.

**Comment by mflatt** [comment]
Maybe we should add `if` and `getenv` to `#lang info`? Then
```
 (define test-omit-paths (if (getenv "PLT_PKG_BUILD_SERVICE") 'all '()))
```
would work.

**Comment by samth** [comment]
`if` seems perfectly safe, but can `getenv` be used in bad ways?

**Comment by mflatt** [comment]
@samth I can't think of a way, but maybe you have something in mind?

**Comment by samth** [comment]
@mflatt Thinking about it more, if you have secrets in environment variables (which is a reasonably common thing to do), this can leak them to the `info.rkt` file. I'm not sure what it can do with it, though.

**Comment by mflatt** [comment]
@samth How about allowing `getenv` but also an environment variable that says which environment variables are accessible to "info.rkt" files? The `get-info` function can prune the environment to just those variables that are to be visible to "info.rkt" files.

**Comment by samth** [comment]
That seems reasonable. What about just a prefix on the allowed environment variable names?

**Comment by mflatt** [comment]
I considered that, but besides the ugliness of a prefix convention, I think this is a case where it makes more sense to give the entity defining an environment more complete control over what to expose.

**Comment by samth** [comment]
Sounds sensible to me.

**Comment by mflatt** [comment]
Done, but since allowing `getenv` is an extension to `info`, I'm not sure how soon the Rosette package will be able to take advantage of it (and still run on relevant Racket versions).

**Comment by emina** [comment]
Awesome, thank you @mflatt!  I'm happy to wait for the next Racket release (and also pick up Robby's patch for Scribble footnotes).

**Comment by jamesbornholt** [comment]
Thanks again for the fix! Now that 6.6 is out, we updated Rosette's info.rkt to check `PLT_PKG_BUILD_SERVICE` and not run the tests if it's set. But the pkg-build server doesn't seem to set `PLT_INFO_ALLOW_VARS`, so the tests are still trying to run: http://pkg-build.racket-lang.org/server/built/test-fail/rosette.txt

The `PLT_INFO_ALLOW_VARS` variable is set correctly on the snapshot(?) pkg-build server and the tests don't run, as expected: https://plt.eecs.northwestern.edu/pkg-build/server/built/test-success/rosette.txt

Is it possible the two servers are out of sync somehow?
